File: android/app/src/main/java/com/gastosapp/AnalyticsActivity.java
package com.gastosapp;

import android.graphics.Color;
import android.os.Bundle;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import androidx.lifecycle.ViewModelProvider;

import com.gastosapp.data.local.ExpenseEntity;
import com.gastosapp.viewmodel.ExpenseViewModel;
import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import com.github.mikephil.charting.formatter.IndexAxisValueFormatter;

import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class AnalyticsActivity extends AppCompatActivity {

    private TextView tvCurrentMonthTotal;
    private LineChart lineChart;
    private PieChart pieChart;
    private ExpenseViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_analytics);

        viewModel = new ViewModelProvider(this).get(ExpenseViewModel.class);

        initViews();
        observeViewModel();
    }

    private void initViews() {
        ImageButton btnBack = findViewById(R.id.btnBack);
        tvCurrentMonthTotal = findViewById(R.id.tvCurrentMonthTotal);
        lineChart = findViewById(R.id.lineChart);
        pieChart = findViewById(R.id.pieChart);

        btnBack.setOnClickListener(v -> finish());
    }

    private void observeViewModel() {
        viewModel.getTotalMonthlyExpenses().observe(this, total -> {
            if (total != null) {
                tvCurrentMonthTotal.setText(formatCurrency(total));
            } else {
                tvCurrentMonthTotal.setText(formatCurrency(0.0));
            }
        });

        viewModel.getAllExpenses().observe(this, expenses -> {
            if (expenses != null) {
                setupLineChart(expenses);
                setupPieChart(expenses);
            }
        });
    }

    private void setupLineChart(List<ExpenseEntity> expenses) {
        List<Entry> entries = new ArrayList<>();

        // Grouping logic for line chart (simplified: last 4 months or similar)
        // Here we just use mock historical data + current month real data for now
        // to keep it consistent with the previous logic but using real data for
        // "current"

        double currentMonthTotal = 0;
        if (expenses != null) {
            // Filter or calculate current month from expenses if needed,
            // but we already have getTotalMonthlyExpenses() for the text view.
            // For the chart, we iterate:
            for (ExpenseEntity e : expenses) {
                // Here you would normally filter by month.
                // Using a simplified logic for demonstration:
                currentMonthTotal += e.getAmount();
            }
        }

        entries.add(new Entry(0, 1500));
        entries.add(new Entry(1, 2400));
        entries.add(new Entry(2, 1800));
        entries.add(new Entry(3, (float) currentMonthTotal));

        LineDataSet dataSet = new LineDataSet(entries, "Gastos Mensuales");
        dataSet.setColor(Color.parseColor("#3b82f6"));
        dataSet.setValueTextColor(Color.GRAY);
        dataSet.setLineWidth(2f);
        dataSet.setCircleColor(Color.parseColor("#3b82f6"));
        dataSet.setDrawFilled(true);
        dataSet.setFillAlpha(50);
        dataSet.setFillColor(Color.parseColor("#3b82f6"));

        LineData lineData = new LineData(dataSet);
        lineChart.setData(lineData);

        String[] months = new String[] { "Ene", "Feb", "Mar", "Abr" };
        XAxis xAxis = lineChart.getXAxis();
        xAxis.setValueFormatter(new IndexAxisValueFormatter(months));
        xAxis.setPosition(XAxis.XAxisPosition.BOTTOM);
        xAxis.setGranularity(1f);
        xAxis.setDrawGridLines(false);

        lineChart.getAxisRight().setEnabled(false);
        lineChart.getDescription().setEnabled(false);
        lineChart.animateX(1000);
        lineChart.invalidate();
    }

    private void setupPieChart(List<ExpenseEntity> expenses) {
        Map<String, Float> categoryTotals = new HashMap<>();

        for (ExpenseEntity expense : expenses) {
            String category = expense.getCategory();
            float amount = (float) expense.getAmount();
            categoryTotals.put(category, categoryTotals.getOrDefault(category, 0f) + amount);
        }

        List<PieEntry> pieEntries = new ArrayList<>();
        for (Map.Entry<String, Float> entry : categoryTotals.entrySet()) {
            pieEntries.add(new PieEntry(entry.getValue(), entry.getKey()));
        }

        PieDataSet dataSet = new PieDataSet(pieEntries, "");
        dataSet.setColors(new int[] {
                Color.parseColor("#10b981"),
                Color.parseColor("#3b82f6"),
                Color.parseColor("#8b5cf6"),
                Color.parseColor("#ef4444"),
                Color.parseColor("#eab308"),
                Color.parseColor("#6b7280")
        });
        dataSet.setValueTextSize(12f);
        dataSet.setValueTextColor(Color.WHITE);

        PieData pieData = new PieData(dataSet);
        pieChart.setData(pieData);
        pieChart.getDescription().setEnabled(false);
        pieChart.setHoleRadius(50f);
        pieChart.setTransparentCircleRadius(55f);
        pieChart.animateY(1000);
        pieChart.invalidate();
    }

    private String formatCurrency(double amount) {
        NumberFormat format = NumberFormat.getCurrencyInstance(new Locale("es", "MX"));
        return format.format(amount);
    }
}

File: android/app/src/main/kotlin/com/gastosapp/AnalyticsActivity.kt
package com.gastosapp

import android.graphics.Color
import android.os.Bundle
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import com.gastosapp.data.local.ExpenseEntity
import com.gastosapp.databinding.ActivityAnalyticsBinding
import com.gastosapp.viewmodel.ExpenseViewModel
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.data.*
import com.github.mikephil.charting.formatter.IndexAxisValueFormatter
import dagger.hilt.android.AndroidEntryPoint
import java.text.NumberFormat
import java.util.*

@AndroidEntryPoint
class AnalyticsActivity : AppCompatActivity() {

    private lateinit var binding: ActivityAnalyticsBinding
    private val viewModel: ExpenseViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAnalyticsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        initViews()
        observeViewModel()
    }

    private fun initViews() {
        binding.btnBack.setOnClickListener { finish() }
    }

    private fun observeViewModel() {
        viewModel.getTotalMonthlyExpenses().observe(this) { total ->
            binding.tvCurrentMonthTotal.text = formatCurrency(total ?: 0.0)
        }

        viewModel.getAllExpenses().observe(this) { expenses ->
            if (expenses != null) {
                setupLineChart(expenses)
                setupPieChart(expenses)
            }
        }
    }

    private fun setupLineChart(expenses: List<ExpenseEntity>) {
        val entries = ArrayList<Entry>()

        // Simplified grouping logic
        val currentMonthTotal = expenses.sumOf { it.amount }

        entries.add(Entry(0f, 1500f))
        entries.add(Entry(1f, 2400f))
        entries.add(Entry(2f, 1800f))
        entries.add(Entry(3f, currentMonthTotal.toFloat()))

        val dataSet = LineDataSet(entries, "Gastos Mensuales").apply {
            color = Color.parseColor("#3b82f6")
            valueTextColor = Color.GRAY
            lineWidth = 2f
            setCircleColor(Color.parseColor("#3b82f6"))
            setDrawFilled(true)
            fillAlpha = 50
            fillColor = Color.parseColor("#3b82f6")
        }

        binding.lineChart.apply {
            data = LineData(dataSet)
            val months = arrayOf("Ene", "Feb", "Mar", "Abr")
            xAxis.apply {
                valueFormatter = IndexAxisValueFormatter(months)
                position = XAxis.XAxisPosition.BOTTOM
                granularity = 1f
                setDrawGridLines(false)
            }
            axisRight.isEnabled = false
            description.isEnabled = false
            animateX(1000)
            invalidate()
        }
    }

    private fun setupPieChart(expenses: List<ExpenseEntity>) {
        val categoryTotals = HashMap<String, Float>()

        for (expense in expenses) {
            val category = expense.category
            val amount = expense.amount.toFloat()
            categoryTotals[category] = (categoryTotals[category] ?: 0f) + amount
        }

        val pieEntries = ArrayList<PieEntry>()
        for ((category, total) in categoryTotals) {
            pieEntries.add(PieEntry(total, category))
        }

        val dataSet = PieDataSet(pieEntries, "").apply {
            colors = intArrayOf(
                Color.parseColor("#10b981"),
                Color.parseColor("#3b82f6"),
                Color.parseColor("#8b5cf6"),
                Color.parseColor("#ef4444"),
                Color.parseColor("#eab308"),
                Color.parseColor("#6b7280")
            ).toList()
            valueTextSize = 12f
            valueTextColor = Color.WHITE
        }

        binding.pieChart.apply {
            data = PieData(dataSet)
            description.isEnabled = false
            holeRadius = 50f
            transparentCircleRadius = 55f
            animateY(1000)
            invalidate()
        }
    }

    private fun formatCurrency(amount: Double): String {
        val format = NumberFormat.getCurrencyInstance(Locale("es", "MX"))
        return format.format(amount)
    }
}
